toolkit/tools/pkg/licensecheck/licensecheck.go

```diff
@@ -0,0 +1,411 @@
+	logger.Log.Infof("Searching rpms...")
+
+	// Wait for all the workers to finish, updating the progress as results come in
+	numProcessed := 0
+	lastReportPercent := 0
+	for range rpmsToSearchPaths {
+		result := <-resultsChannel
+		if result.err != nil {
+			// Signal the workers to stop if there is an error
+			err = fmt.Errorf("failed to search srpm:\n%w", result.err)
+			close(cancel)
+			return
+		}
+		numProcessed++
+		percentProcessed := (numProcessed * 100) / len(rpmsToSearchPaths)
+		if percentProcessed-lastReportPercent >= searchReportIntervalPercent {
+			logger.Log.Infof("Checked %d/%d rpms (%d%%)", numProcessed, len(rpmsToSearchPaths), percentProcessed)
+			lastReportPercent = percentProcessed
+		}
+		results = append(results, result)
+	}
+	return
+}
+
+// findRpmPaths walks the chroots's mount directory to find all *.rpm files. The paths are returned relative to the
+// chroot's root.
+func (s *LicenseChecker) findRpmPaths() (foundSrpmPaths []string, err error) {
+	const rpmExtension = ".rpm"
+	err = filepath.Walk(s.simpleToolChroot.ChrootRelativeMountDir(), func(path string, info os.FileInfo, walkErr error) error {
+		if walkErr != nil {
+			return walkErr
+		}
+		if info.IsDir() {
+			return nil
+		}
+		if !strings.HasSuffix(path, rpmExtension) {
+			return nil
+		}
+
+		foundSrpmPaths = append(foundSrpmPaths, path)
+		return nil
+	})
+	if err != nil {
+		err = fmt.Errorf("failed to walk directory:\n%w", err)
+		return nil, err
+	}
+	return foundSrpmPaths, nil
+}
+
+// queueWorkers queues up workers to search the RPMs in parallel. Each worker will wait on the jobSemaphore before starting.
+func (l *LicenseChecker) queueWorkers(rpmsToSearchPaths []string, resultsChannel chan LicenseCheckResult, cancel chan struct{}) {
+	for _, rpmPath := range rpmsToSearchPaths {
+		// Wait for the semaphore, or allow cancel before running
+		select {
+		case l.jobSemaphore <- struct{}{}:
+		case <-cancel:
+			return
+		}
+		go func(rpmPath string) {
+			defer func() {
+				<-l.jobSemaphore
+			}()
+
+			logger.Log.Debugf("Searching (%s)", filepath.Base(rpmPath))
+			searchResult, err := CheckRpmLicenses(rpmPath, l.distTag, l.licenseNames, l.exceptions)
+			logger.Log.Debugf("Finished searching (%s)", filepath.Base(rpmPath))
+			if err != nil {
+				logger.Log.Errorf("License check worker failed with error: %v", err)
+				resultsChannel <- LicenseCheckResult{err: err}
+				return
+			}
+			resultsChannel <- searchResult
+		}(rpmPath)
+	}
+}
+
+// CheckRpmLicenses checks the licenses of an RPM at the given path. It returns result struct holding all the license
+// issues found. This function will use the host's macros to query the RPM so it is expected to be called in a chroot.
+// - rpmPath: The path to the RPM to check relative to the chroot's root.
+func CheckRpmLicenses(rpmPath, distTag string, licenseNames LicenseNames, exceptions LicenseExceptions) (result LicenseCheckResult, err error) {
```

Comment:
Why a public function? This makes this module be in a "split" position, where some functionality is exposed through the `LicenseChecker` struct and some through separate public functions. It seems like `CheckRpmLicenses` is only used internally by the `LicenseChecker` anyway, so I'd suggest keeping it private at least for now. If we learn in the future, that it's useful to expose this functionality publicly, that's going to be easy to do. Doing it upfront now may cause unnecessary confusion about how to use this module.

Reply:
This will be needed by the build workers, this is just part one of the changes. The build worker has no need to stand up a full chroot object, it just needs to scan some files.

Go packaging best practices is to include all related functionality in a single package, don't over split.

Reply:
Having used the API for all its intended purposes, using an existing chroot was too much trouble, easier to just use the dedicated chroot using `New()`. Will make this private.